apps/src/weblab2/Weblab2View.tsx

```diff
@@ -148,7 +148,10 @@ const Weblab2View = () => {
 
   const setProject = (newProject: MultiFileSource) => {
     setCurrentProject(newProject);
-    if (Lab2Registry.getInstance().getProjectManager()) {
+    if (
+      Lab2Registry.getInstance().getProjectManager() &&
+      newProject !== currentProject
```

Comment:
if we blindly save even if there's no change, it can overwrite if you navigate too fast. 

This is kinda a hack solution - it'd be better for CDOIDE not to fire off the additional noisy note, and it'd also be better if the project manager's save could somehow take into account if nothing's changing (but that's probably asking too much since it doesn't maintain state otherwise).

Reply:
The project manager actually does check if nothing has changed, and does not save in that case ([link](https://github.com/code-dot-org/code-dot-org/blob/457197fcbc2d64dc1bb2c3973ef89d7dc0fd5ed1/apps/src/lab2/projects/ProjectManager.ts#L287)). Is the problem here that the channel id has changed but we are sending a late update with the code from the previous channel id?

Reply:
The problem was that the `setProject` function wasn't memoized.

In turn, when you navigated it'd create a new `setProject` function which'd fall into the IDE and trigger the effect to claim that the `project` had changed, even though it hadn't in this case. With a memoized function, it behaves as expected.

Further, though, this is a bigger issue beyond this code - because the project manager's `save` method only takes _data_ to save and not the channelID to save it to (I think I'm using the right terms), it's susceptible to this type of issue -

basically, if the user navigates old data from the prior level may still be loaded into components. If something is triggered which would cause a save, it might push forward the stale data to the project manager, which would then overwrite the _new_ channelID with the prior channel's data. It's assuming that all data that currently exists on the page is new, but if it hasn't been refreshed yet it's stale, and if that stale data triggers a save, it'd overwrite.

As it did in this case since the function wasn't memoized. This should be flagged as a future fix to circle back to.

Reply:
I made a follow up task to look into the caching idea I came up with: https://codedotorg.atlassian.net/browse/CT-500